scheduler/scheduler.cpp

```diff
@@ -742,26 +741,18 @@ static CompileServer *pick_server(Job *job)
                 " client count: " << cs->clientCount() << endl;
 #endif
 
-        if ((cs->lastCompiledJobs().size() == 0) && (cs->jobList().size() == 0) && cs->maxJobs()) {
-            /* Make all servers compile a job at least once, so we'll get an
-               idea about their speed.  */
-            if (!envs_match(cs, job).empty()) {
-                best = cs;
-                matches++;
-            } else {
-                // if there is one server that already got the environment and one that
-                // hasn't compiled at all, pick the one with environment first
-                bestui = cs;
-            }
-
-            break;
+        // Some portion of the selection will go to a host that has not been selected
+        // in a while so we can maintain reasonably up-to-date statistics. The greater
+        // the weight, the less likely this is to happen. If weight == UINT8_MAX, we
+        // *always* want whatever is deemed to be the fastest server.
+        uint8_t weight_limit = std::numeric_limits<uint8_t>::max() - weight;
+        uint8_t weight_factor = 0;
+        if (weight_limit > 0) {
+            weight_factor = std::numeric_limits<uint8_t>::max() / weight_limit;
         }
 
-        /* Distribute 5% of our jobs to servers which haven't been picked in a
-           long time. This gives us a chance to adjust the server speed rating,
-           which may change due to external influences out of our control. */
-        if (!cs->lastPickedId() ||
-            ((job->id() - cs->lastPickedId()) > (20 * css.size()))) {
+        if (weight_factor > 0 && (!cs->lastPickedId() ||
+            ((job->id() - cs->lastPickedId()) > (weight_factor * eligible.size())))) {
```

Comment:
it's not easily identifiable what exactly this condition is doing, could you add a comment explaining it. For example, what do job ids have to do with the number of eligible servers?

Reply:
Job id's are a monotonically increasing sequence. By looking for the last job id a server did, I can tell when it was last used. This is how another (unchanged) part of the code does something similar.